controllers/disruption_controller.go

```diff
@@ -343,91 +531,128 @@ func (r *DisruptionReconciler) updateInjectionStatus(instance *chaosv1beta1.Disr
 
 // startInjection creates non-existing chaos pod for the given disruption
 func (r *DisruptionReconciler) startInjection(instance *chaosv1beta1.Disruption) error {
-	var err error
-
 	if len(instance.Status.Targets) > 0 {
 		r.log.Infow("starting targets injection", "targets", instance.Status.Targets)
 	}
 
+	// chaosPodsMap will be used to check if a target has an existing chaos pod already or not
+	chaosPodsMap := map[string]corev1.Pod{}
+
+	chaosPods, err := r.getChaosPods(instance, nil)
+	if err != nil {
+		return fmt.Errorf("error getting chaos pods: %w", err)
+	}
+
+	for _, chaosPod := range chaosPods {
+		if !utils.Contains(instance.Status.Targets, chaosPod.Labels[chaostypes.TargetLabel]) {
+			r.deleteChaosPod(instance, chaosPod)
+		} else {
+			chaosPodsMap[chaosPod.Labels[chaostypes.TargetLabel]+"-"+chaosPod.Labels[chaostypes.DisruptionKindLabel]] = chaosPod
+		}
+	}
+
+	// iterate through target + existing disruption kind -- to ensure all chaos pods exist
 	for _, target := range instance.Status.Targets {
-		targetNodeName := ""
-		targetContainerIDs := []string{}
-		targetPodIP := ""
-		chaosPods := []*corev1.Pod{}
-
-		// retrieve target
-		switch instance.Spec.Level {
-		case chaostypes.DisruptionLevelUnspecified, chaostypes.DisruptionLevelPod:
-			pod := corev1.Pod{}
-
-			if err := r.Get(context.Background(), types.NamespacedName{Namespace: instance.Namespace, Name: target}, &pod); err != nil {
-				return fmt.Errorf("error getting target to inject: %w", err)
+		for _, kind := range chaostypes.DisruptionKindNames {
+			if subspec := instance.Spec.DisruptionKindPicker(kind); reflect.ValueOf(subspec).IsNil() {
+				continue
 			}
 
-			targetNodeName = pod.Spec.NodeName
+			if _, ok := chaosPodsMap[target+"-"+kind.String()]; ok {
+				continue
+			}
 
-			// get IDs of targeted containers or all containers
-			targetContainerIDs, err = getContainerIDs(&pod, instance.Spec.Containers)
-			if err != nil {
-				return fmt.Errorf("error getting target pod container ID: %w", err)
+			if err = r.createChaosPods(instance, target); err != nil {
+				return fmt.Errorf("error creating chaos pods: %w", err)
 			}
 
-			// get IP of targeted pod
-			targetPodIP = pod.Status.PodIP
-		case chaostypes.DisruptionLevelNode:
-			targetNodeName = target
+			break
```

Comment:
Why `break` here?

Reply:
The scenario is reached when there are multiple kinds on the disruption, and multiple chaos pods for one target. In the event a chaos pod is missing, `r.createChaosPods` is launched, which will attempt to create all chaos pods for a target (and cancel the creation of the already existing ones with a debug-only line). It shouldn't run again for this target.
I think this logic is fine, we can discuss it internally elsewhere.
--
If deemed problematic, this should be adjusted later in another PR; chaos pods could indeed be "created" individually for each target+disruption kind and not bundled like this -- but that's not the primary focus of this PR that is already a heavy bit of change.